File: problems/763/solution.go
package main
func partitionLabels(S string) []int {
    lastSeen, ans := make(map[rune]int), []int{}
    
    for i, ch := range S {
        lastSeen[ch] = i
    }
    
    partStart, partEnd := 0, 0

    for i, ch := range S {
        partEnd = max(partEnd, lastSeen[ch])
        
        if i == partEnd {
            ans = append(ans, partEnd - partStart + 1)
            partStart = i + 1
        }
    }
    
    return ans
}

func max(a, b int) int {
    if a > b {
        return a
    }
    
    return b
}

File: problems/763/solution.cpp
class Solution {
public:
    vector<int> partitionLabels(string S) {
        vector<int> lastSeen(26);
        vector<int> ans;
        
        for (int i = 0; i < S.size(); i++) {
            lastSeen[S[i] - 'a'] = i;
        }
        
        int partStart = 0, partEnd = 0;
        
        for (int i = 0; i < S.size(); i++) {
            partEnd = max(partEnd, lastSeen[S[i] - 'a']);
            
            if (partEnd == i) {
                ans.push_back(partEnd - partStart + 1);
                partStart = i + 1;
            }
        }
        
        return ans;
    }
};